partiql-common/src/metadata.rs

```diff
@@ -0,0 +1,366 @@
+        let value = self.inner.get(key);
+        if let Some(PartiqlMetaValue::Array(v)) = value {
+            Some(v.clone())
+        } else {
+            None
+        }
+    }
+
+    pub fn bool_value(&self, key: &str) -> Option<bool> {
+        let value = self.inner.get(key);
+        if let Some(PartiqlMetaValue::Bool(v)) = value {
+            Some(*v)
+        } else {
+            None
+        }
+    }
+
+    pub fn f32_value(&self, key: &str) -> Option<f32> {
+        let value = self.inner.get(key);
+        if let Some(PartiqlMetaValue::Float32(v)) = value {
+            Some(*v)
+        } else {
+            None
+        }
+    }
+
+    pub fn f64_value(&self, key: &str) -> Option<f64> {
+        let value = self.inner.get(key);
+        if let Some(PartiqlMetaValue::Float64(v)) = value {
+            Some(*v)
+        } else {
+            None
+        }
+    }
+
+    pub fn decimal_value(&self, key: &str) -> Option<Decimal> {
+        let value = self.inner.get(key);
+        if let Some(PartiqlMetaValue::Decimal(v)) = value {
+            Some(*v)
+        } else {
+            None
+        }
+    }
+
+    pub fn i32_value(&self, key: &str) -> Option<i32> {
+        let value = self.inner.get(key);
+        if let Some(PartiqlMetaValue::Int32(v)) = value {
+            Some(*v)
+        } else {
+            None
+        }
+    }
+
+    pub fn i64_value(&self, key: &str) -> Option<i64> {
+        let value = self.inner.get(key);
+        if let Some(PartiqlMetaValue::Int64(v)) = value {
+            Some(*v)
+        } else {
+            None
+        }
+    }
+
+    pub fn map_value(&self, key: &str) -> Option<PartiqlMetadata<T>> {
+        let value = self.inner.get(key);
+        if let Some(PartiqlMetaValue::Map(v)) = value {
+            Some(v.clone())
+        } else {
+            None
+        }
+    }
+
+    pub fn string_value(&self, key: &str) -> Option<String> {
+        let value = self.inner.get(key);
+        if let Some(PartiqlMetaValue::String(v)) = value {
+            Some(v.clone())
+        } else {
+            None
+        }
+    }
+}
+
+impl<T> Default for PartiqlMetadata<T>
+where
+    T: Eq + Clone + Hash + Borrow<str>,
+{
+    fn default() -> Self {
+        Self {
+            inner: IndexMap::new(),
+        }
+    }
+}
+
+impl<T> FromIterator<(T, PartiqlMetaValue<T>)> for PartiqlMetadata<T>
+where
+    T: Eq + Clone + Hash + Borrow<str>,
+{
+    fn from_iter<I: IntoIterator<Item = (T, PartiqlMetaValue<T>)>>(iter: I) -> Self {
+        let inner = iter.into_iter().collect();
+        Self { inner }
+    }
+}
+
+impl<T> IntoIterator for PartiqlMetadata<T>
+where
+    T: Eq + Clone + Hash + Borrow<str>,
+{
+    type Item = (T, PartiqlMetaValue<T>);
+    type IntoIter = indexmap::map::IntoIter<T, PartiqlMetaValue<T>>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.inner.into_iter()
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum PartiqlMetaValue<T>
```

Comment:
May want to `#[allow(dead_code)]` on this, since the variants may not all be used in places that partiql-lang-rust can see.

Reply:
done